entry/src/block_component.rs

```diff
@@ -0,0 +1,3437 @@
+/// ### BlockComponent with BlockMarker
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Entry Count = 0              (8 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ Marker Version               (2 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ Marker Data               (variable)    │
+/// └─────────────────────────────────────────┘
+/// ```
+///
+/// ### BlockMarkerV1 Layout
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Variant ID                   (1 byte)   │
+/// ├─────────────────────────────────────────┤
+/// │ Byte Length                  (2 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ Variant Data              (variable)    │
+/// └─────────────────────────────────────────┘
+/// ```
+///
+/// ### BlockFooterV1 Layout
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Producer Time Nanos          (8 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ User Agent Length            (1 byte)   │
+/// ├─────────────────────────────────────────┤
+/// │ User Agent Bytes          (0-255 bytes) │
+/// └─────────────────────────────────────────┘
+/// ```
+///
+/// ### BlockHeaderV1 Layout
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Parent Slot                  (8 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ Parent Block ID             (32 bytes)  │
+/// └─────────────────────────────────────────┘
+/// ```
+///
+/// ### UpdateParentV1 Layout
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Parent Slot                  (8 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ Parent Block ID             (32 bytes)  │
+/// └─────────────────────────────────────────┘
+/// ```
+use {
+    crate::entry::Entry,
+    serde::{
+        de::{self, Visitor},
+        Deserialize, Deserializer, Serialize, Serializer,
+    },
+    solana_clock::Slot,
+    solana_hash::Hash,
+    std::{error::Error, fmt},
+};
+
+/// Error types for block component operations.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum BlockComponentError {
+    /// Data is too short for the expected format
+    InsufficientData,
+    /// Entry count exceeds the maximum allowed
+    TooManyEntries { count: usize, max: usize },
+    /// EntryBatch is empty when it shouldn't be
+    EmptyEntryBatch,
+    /// Unknown variant identifier
+    UnknownVariant { variant_type: String, id: u8 },
+    /// Unsupported version number
+    UnsupportedVersion { version: u16 },
+    /// Data length conversion failed (e.g., usize to u16)
+    DataLengthOverflow,
+    /// Cursor position exceeded data boundary
+    CursorOutOfBounds,
+    /// BlockComponent cannot have both entry batch and marker data
+    MixedData,
+    /// Serialization failed
+    SerializationFailed(String),
+    /// Deserialization failed
+    DeserializationFailed(String),
+}
+
+impl fmt::Display for BlockComponentError {
```

Comment:
Consider deriving `Error` for this type and then you do not need to implement `Display`.  See consensus_pool.rs for an example.